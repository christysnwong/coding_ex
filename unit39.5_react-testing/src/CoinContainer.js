import React, { useState } from "react";
import Coin from "./Coin";

function CoinContainer(props) {

  // customize the # of coins, can set more than 1 coin
  let numCoins = 4;

  const [coins, setCoins] = useState([]);
  const [headCount, setHeadCount] = useState(0);
  const [tailCount, setTailCount] = useState(0);

  const flipCoins = () => {
    let newCoins = [];

    for (let i = 0; i < numCoins; i++) {
      const randIdx = Math.floor(Math.random() * 2);
      let newCoin = props.coins[randIdx];

      if (newCoin.side === "head") {
        setHeadCount((headCount) => headCount + 1);
      } else {
        setTailCount((tailCount) => tailCount + 1);
      }

      newCoins.push(newCoin);
    }
    
    setCoins(newCoins);
    
  };

  const reset = () => {
    setCoins([]);
    setHeadCount(0);
    setTailCount(0);
  }
  

  return (
    <div className="CoinContainer">
      <h2>Let's flip coin(s)!</h2>

      {coins.map((c, idx) => (
        <Coin key={idx} side={c.side} src={c.src} />
      ))}

      <p>
        <button onClick={flipCoins}>Flip!</button>
        <button onClick={reset}>Reset</button>
      </p>
      <p>
        Out of {(headCount + tailCount) / numCoins} flip(s),
        there are {headCount} heads and {tailCount} tails.
      </p>
    </div>
  );
}

CoinContainer.defaultProps = {
  coins: [
    {
      side: "head",
      src: "https://www.coinsunlimited.ca/image/cache/wp/cp/data/decimal-coins-canada/rolls/2022/2022-canadian-1-dollar-common-loon-loonie-coin-2-800x800.webp",
    },
    {
      side: "tail",
      src: "https://www.coinsunlimited.ca/image/cache/wp/cp/data/decimal-coins-canada/rolls/2022/2022-canadian-1-dollar-common-loon-loonie-coin-1-800x800.webp",
    },
  ],
};

export default CoinContainer;