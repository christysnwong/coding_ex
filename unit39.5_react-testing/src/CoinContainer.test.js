import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import CoinContainer from "./CoinContainer";

// this test is written for 4 coins flip

beforeEach(function () {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75)
    .mockReturnValue(0.25);
});

afterEach(function () {
  Math.random.mockRestore();
});

it("renders without crashing", function () {
  render(<CoinContainer />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<CoinContainer />);
  expect(asFragment()).toMatchSnapshot();
});

it("doesn't show a coin on page load", function() {
    const { queryByAltText } = render(<CoinContainer />);

    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();

});


it("counts heads properly", function () {
  const { getByText, queryAllByAltText } = render(<CoinContainer />);

  const button = getByText("Flip!");
  fireEvent.click(button);

  expect(queryAllByAltText("head")).toHaveLength(4);
  expect(queryAllByAltText("tail")).toHaveLength(0);
  expect(getByText("Out of 1 flip(s), there are 4 heads and 0 tails.")).toBeInTheDocument();
});

it("counts tails properly", function () {
  const { getByText, queryAllByAltText } = render(<CoinContainer />);

  const button = getByText("Flip!");
  fireEvent.click(button);
  fireEvent.click(button);

  screen.logTestingPlaygroundURL();

  expect(queryAllByAltText("head")).toHaveLength(3);
  expect(queryAllByAltText("tail")).toHaveLength(1);
  expect(
    getByText("Out of 2 flip(s), there are 7 heads and 1 tails.")
  ).toBeInTheDocument();
});