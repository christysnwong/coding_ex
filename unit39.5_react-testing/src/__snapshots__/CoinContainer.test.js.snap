// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinContainer"
  >
    <h2>
      Let's flip coin(s)!
    </h2>
    <p>
      <button>
        Flip!
      </button>
      <button>
        Reset
      </button>
    </p>
    <p>
      Out of 0 flip(s), there are 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
